### ISSUE TITLE
Websocket / Partial Write / lws_buflist_append_segment: buflist reached sanity limit

### ISSUE THREAD

**Issue by dev31337** [warmcat/libwebsockets#1964]
Hi,

For a websocket connection,

lws_write() used on a large packet (8K) returns with partial write.

The documentation says that in this case:

"LWS will buffer the remainder automatically, and send it out autonomously."

So basically we ignore the partial write and we keep writing the next packet at the next LWS_CALLBACK_SERVER_WRITEABLE event.

If we do this the lws_write() starts returning "0" then "-1" then ultimately the error "lws_buflist_append_segment: buflist reached sanity limit".

Am I missing something here ?

Why would lws trigger a LWS_CALLBACK_SERVER_WRITEABLE  event if it can not write any data ?

Regards.

**Comment by lws-team** [comment]
> So basically we ignore the partial write and we keep writing the next packet at the next LWS_CALLBACK_SERVER_WRITEABLE event.

Right.

> Why would lws trigger a LWS_CALLBACK_SERVER_WRITEABLE event if it can not write any data ?

lws just reacts to what the OS tells it, if POLLOUT comes up on the poll or select wait for that fd, we assume we can write something.  But even if partial again, it should not give you another WRITEABLE until it has emptied the output buflist.

This "sanity limit" is the buflist finding that it has 1024 heap buffers allocated... that's actually pretty insane.

Can I suggest you try to adapt a similar minimal example to show the same behaviour (on master)?  If it's broken there provide the patch on the example and I will study it.

**Comment by dev31337** [comment]
I see, I will try to modify a minimal example, basically the idea is to lws_write() 8K each time you get a LWS_CALLBACK_SERVER_WRITEABLE event until it produces the error.

Meanwhile do you know if there is a way to increase the SO_SNDBUF of the ws socket using lws ?

**Comment by dev31337** [comment]
There is definitely a bug in your buffering system, when a packet is buffered the remote does not receive the right complete packet.

I was able to overcome this issue by avoiding triggering your buffering system by looking in your source and doing this:

. Increase the tx_packet_size of the ws lws_protocols to the maximum possible packet size. (this avoid your buffering system to be triggered at write time)
. Do a usleep at each lsw_write to be sure the SO_SNDBUF buffer is never filled. (this avoid your buffering system to be triggered on partial write)

I'm not sure I will be able to modify a minimal example to show you the bug live, I strongly suggest you to modify it yourself if you want to be sure your buffering code is bug free.

I was able to find the time when you created this buffering system and your post suggests that you did not test it:

https://libwebsockets.org/pipermail/libwebsockets/2013-October/004460.html

"I added a patched aimed at comprehensively handling partial sends.  It's 
not tested due to problems reproducing but it should be no worse than 
what was there before this patch"

**Comment by lws-team** [comment]
Hm... in the 7 years since then this code has had a lot of work and changes, most recently a couple of versions ago to use lws_buflists.

>> not tested due to problems reproducing but it should be no worse than
what was there before this patch

I fixed the guy's problem 7 years ago and I should feel bad I couldn't reproduce his problem then 7 years later?  The code from then bears no relation to the code today.

> I'm not sure I will be able to modify a minimal example to show you the bug live, 

Now I can't easily reproduce your problem either.

I get users with such high standards it's a shame they don't contribute some awesome code to show me how it should be done.

**Comment by dev31337** [comment]
I just meant that it seems it wasn't tested and thus could explain the bug. (you seem to be always on the defensive in your different posts, I mean and meant no harm)

If I have trouble making a minimal example is that I have to fight with a tight roadmap and a boss that wants results fast, I will try to make an minimal example on my free time to avoid any issue at work.

Meanwhile I was able to make a fully working workaround by removing O_NONBLOCK on the wsi socket before the lws_write (and restoring it afterwards) to avoid triggering the bug of partial buffering.

It's blocking the main thread until the write is fully done but it's totally fine for me.

**Comment by dev31337** [comment]
I learnt how you code is working on lsi_write() and I was able to overcome the issue by doing probably what your code is expecting.

It seems that your code is expecting that during one LWS_CALLBACK_SERVER_WRITEABLE event you do only one single lswi_write().

This was not the case with my code, I was lswi_writing all the packets in the queue, and this was triggering the issue.

I modified my code to done one single lsi_write and then trigger a lws_callback_on_writable() if there is more to send/write.

Now everything is working perfectly, your buffering is in action and works as it should.
